### PR TITLE
Set up a Linux workflow to work with ARC runners

### PR DESCRIPTION
**Pull request by DanilBaibak**
Set up a Linux workflow to work with ARC runners (rootless DinD).

Changes implemented:
* `linux-jammy-py3.8-gcc11` builds and tests were for some reason picking up wrong ldcxx lib, could not find a proper way to deal with this except from deleting the roge library from the docker container. I want to reiterate on this in a next step;
* S3 bucket to store temporary build assets changed as it is migrating to the new AWS account. To make the migration smooth, I decided to create this as an optional input to the relevant workflows;
* Currently we rely a lot of access permission on what is assigned to the AWS nodes. This is not great from a security standpoint and it is a nightmare to migrate. This is why I am migrating to use OICD as much as possible. This have been implemented already for docker download and upload, s3 access and other parts. The idea is to use the ARC migration to remove the reliance of permissions granted to the node itself;

This change depends on the following changes: 
* Force docker to update credentials: https://github.com/pytorch/test-infra/pull/4991
* Add permissions to role to access ECR: https://github.com/pytorch/ci-infra/commit/acc0154aa0289369411c2f3105a21ce6010a1da4
* Add permissions to the role to access relevant S3 bucket: https://github.com/pytorch/ci-infra/commit/496b0422c360fb5ca98b9de5db0e1d8f749b22cc

## Reasoning for introducing a new `_linux-build-arc.yml`

Old style `runs-on` definition accept a string, new style `runs-on` requires a object in the format:

```
--- old
... 
  runs-on: "linux.2xlarge"
...
--- new
...
  runs-on:
    group: "running-group"
...
```

In other words, to specify a group the format of the yaml needs to be changed. Unfortunately, there is no way to accomplish this change using any trick in the book that I am aware of. This is due to the fact that GH actions yaml are not templatable and support minimal functions / replacements. A few examples that did not work: 
* [`e234f25` (#119544)](https://github.com/pytorch/pytorch/pull/119544/commits/e234f25ba1dca63989582e521779579d9e1760ad#diff-b317d4da565a9e329ccf67e669c2ff1f4d4bc5fb0ffa4d74132545ad66f84339R76)
* [`087de4a` (#119544)](https://github.com/pytorch/pytorch/pull/119544/commits/087de4ad8ba5babc95334d0d65d17293adf7b660#diff-b317d4da565a9e329ccf67e669c2ff1f4d4bc5fb0ffa4d74132545ad66f84339R76)
* [`f03512e` (#119544)](https://github.com/pytorch/pytorch/pull/119544/commits/f03512e344bf46d7d0cd5687061a4a482ee1e9de#diff-b317d4da565a9e329ccf67e669c2ff1f4d4bc5fb0ffa4d74132545ad66f84339R76)
* [`67581fb` (#119544)](https://github.com/pytorch/pytorch/pull/119544/commits/67581fb737d6f2482d53fd344e4fa91ebdb8a029#diff-b317d4da565a9e329ccf67e669c2ff1f4d4bc5fb0ffa4d74132545ad66f84339R76)

### CODE REVIEW

**Review by malfet — CHANGES_REQUESTED** [review]
Would it be possible to add `use_rootless_did` as an option to `_linux_build` workflow, rather than copy-n-paste 75% of the code as `_linux-build-arc`?

**Comment by malfet** [comment]
Sorry, if this is still a work-in-progress PR, do you mind converting it to draft?

File: .ci/docker/build.sh
Comment (seemethere) [inline .ci/docker/build.sh:74]: What does this do? I consider myself a bash expert and I'm a bit confused here.

**Comment by seemethere** [comment]
> Would it be possible to add `use_rootless_did` as an option to `_linux_build` workflow, rather than copy-n-paste 75% of the code as `_linux-build-arc`?

Agree with @malfet here, is there a way to do this without having to have such a split path? Seems like even the docker image stuff might not need a completely separate path / dockerfile?

File: .ci/docker/build.sh
Comment (DanilBaibak) [reply .ci/docker/build.sh]: It does 'shift' but we don't need it anymore.

File: .ci/docker/common/install_user.sh
Comment (zxiiro) [inline .ci/docker/common/install_user.sh:9]: Seems like there's some history I'm missing here that I'd like to understand. I'd like to understand why we need a Jenkins user if we're not using Jenkins?

Comment (jeanschmidt) [reply .ci/docker/common/install_user.sh:9]: One thing that bothers me, we're suppose to be running DID rootless. 

Then we create a jenkins user, set it to the same UID as the runner user, and add it to /etc/sudoers. So we're not doing rootless then...?

Comment (zxiiro) [reply .ci/docker/common/install_user.sh:9]: @jeanschmidt that's correct. If we are adding the runner user to sudoer's and allowing that user to switch to root then we indeed are not doing dind-rootless. We'd need to lockdown the runner user to prevent it from calling sudo if we want to be fully rootless.

File: .github/actions/download-build-artifacts/action.yml
Comment (zxiiro) [inline .github/actions/download-build-artifacts/action.yml:17]: Instead of saying LF bucket should we say "PyTorch" instead?

File: .ci/pytorch/build.sh
Comment (zxiiro) [inline .ci/pytorch/build.sh:260]: We could save 2 lines of code by making this do `sudo chown -R jenkins ../workspace`

File: .github/workflows/_linux-build.yml
Comment (zxiiro) [inline .github/workflows/_linux-build.yml:92]: I assume this will be updated before merge?
(Same with the other instances below)

File: .github/workflows/pull.yml
Comment (zxiiro) [inline .github/workflows/pull.yml:14]: Looks like an accidental space was added here.

File: .github/workflows/pull.yml
Comment (zxiiro) [inline .github/workflows/pull.yml:37]: Don't forget to update this before merge.

**Review by zxiiro — CHANGES_REQUESTED** [review]

File: .ci/pytorch/build.sh
Comment (atalman) [inline .ci/pytorch/build.sh:10]: Please search for existing issue or create a new issue for us in ``pytorch/pytorch`` with description of the error you are seeing.
Also worthwhile checking the anaconda issues here:
https://github.com/ContinuumIO/anaconda-issues i

File: .github/actions/setup-linux/action.yml
Comment (atalman) [inline .github/actions/setup-linux/action.yml:30]: Perhaps we can add : ``! -e /.inarc`` as ENV variable and then this step could be skipped:
``
if: ${{ ! IN_ARC }}
``

**Review by atalman — CHANGES_REQUESTED** [review]
Please fix lint and address comments

File: .github/actions/setup-linux/action.yml
Comment (jeanschmidt) [reply .github/actions/setup-linux/action.yml:30]: good point, using `if: ${{ hashFiles('/.inarc') == '' }}` seems even better, as syntax beautifulness is concerned

File: .ci/pytorch/build.sh
Comment (jeanschmidt) [reply .ci/pytorch/build.sh:10]: Opened a issue with very clear description of the origin of the problem https://github.com/pytorch/pytorch/issues/121796

TLDR: no, the problem is not related to ARC, and it does not necessarily happens on ARC runners. So doing this only when running on ARC won't help. I tried it in multiple ways :(

File: .github/workflows/pull.yml
Comment (zxiiro) [reply .github/workflows/pull.yml:14]: Looks like the extra space is still here.

**Review by zxiiro — APPROVED** [review]
Approved. One comment on a very minor detail but I don't think it should hold up this PR.

File: .ci/pytorch/build.sh
Comment (atalman) [reply .ci/pytorch/build.sh:10]: Please aldo list this issue in the comments. We probably don't need to use echo here
Use something like this:
```
# Forces build to rely on system libstdc++ rather then anaconda provided one see https://github.com/pytorch/pytorch/issues/121796
```

File: .github/workflows/_linux-build.yml
Comment (atalman) [inline .github/workflows/_linux-build.yml:190]: We need to check it passes on other linux runners like aarch64.

**Review by atalman — APPROVED** [review]

File: .ci/pytorch/build.sh
Comment (atalman) [inline .ci/pytorch/build.sh:9]: Instead of checking for BUILD_ENVIRONMENT can we perhaps check if its anaconda kernel5+ ?

Comment (jeanschmidt) [reply .ci/pytorch/build.sh:9]: unfortunately no, we would need to check the combination build + test for it to work.

File: .github/workflows/_linux-build.yml
Comment (jeanschmidt) [reply .github/workflows/_linux-build.yml:190]: what labels do I need to include in order to confirm this?

File: .ci/pytorch/test.sh
Comment (malfet) [inline .ci/pytorch/test.sh:29]: This should not longer be needed, as we install the uptodate libstdc++ into our conda environment now

**Comment by jeanschmidt** [comment]
Merged main back to test with changes introduced on https://github.com/pytorch/pytorch/pull/121556

File: .github/workflows/_linux-build.yml
Comment (atalman) [reply .github/workflows/_linux-build.yml:190]: @jeanschmidt ciflow/binaries_wheel to trigger aarch64 wheel builds

File: .ci/docker/build.sh
Comment (malfet) [inline .ci/docker/build.sh:364]: Just curious, why this is needed?

File: .ci/pytorch/build.sh
Comment (malfet) [inline .ci/pytorch/build.sh:253]: What is wrong with doing it for libtorch?

File: .github/actions/setup-linux/action.yml
Comment (malfet) [inline .github/actions/setup-linux/action.yml:37]: I'm not a shell expert, but there must be a more compact way of doing it, for example
```suggestion
        echo "IN_ARC_RUNNER=$(if [ -f /.inarc ]; then echo true; else echo false; fi)"  >> $GITHUB_OUTPUT
```

File: .github/workflows/_linux-build.yml
Comment (malfet) [reply .github/workflows/_linux-build.yml:190]: I don't know what you guys are trying to do here, but this is very/very/very wrong. One should not assume that they have `sudo` privileges while running the actions

File: .ci/pytorch/build.sh
Comment (malfet) [inline .ci/pytorch/build.sh:244]: Should this be part of docker build (I remember adding this one in the past, see https://github.com/pytorch/pytorch/issues/78659#issuecomment-1144107327 )

**Review by malfet — COMMENTED** [review]
Do you guys mind changing this one into stack of changes? As some of them are completely beign (like adding `--no-same-owner` or safe directory or adding `S3_BUCKET` as parameter name) but some needs further discussion (like adding `sudo` to artifact unzip)

**Review by malfet — CHANGES_REQUESTED** [review]
It would be good to outline in the PR description, what's the difference between `_linux-build` and `_linux-build-arc` and why one needs a new .yml file instead of reusing/extending existing `_linux-build.yml`?

**Comment by jeanschmidt** [comment]
@malfet 

:(

Is this description a blocker to move forward? I believe if this is the only issue you could've approve it and left a message to update before doing so.

File: .ci/docker/build.sh
Comment (jeanschmidt) [reply .ci/docker/build.sh:364]: Danylo added it, I did not remove. I can try without it.

@DanilBaibak what is the reasoning here?

File: .github/actions/setup-linux/action.yml
Comment (jeanschmidt) [reply .github/actions/setup-linux/action.yml:37]: it is the trade off between cleverness and readability. But if the goal is being clever, I believe this is the best way to do it:

```
echo "IN_ARC_RUNNER=$([ -f /.inarc ] && echo true || echo false)"  >> $GITHUB_OUTPUT
```

**Comment by jeanschmidt** [comment]
@malfet Updated description with the reasoning for _linux-build-arc.yml: TLDR I tried for days to do in a single one, but failed miserably. Still open to ideas in how-to.

File: .github/workflows/_linux-build.yml
Comment (jeanschmidt) [reply .github/workflows/_linux-build.yml]: removed, seems that after all the changes made in the arc runner itself, the sudo here was not needed anymore

File: .ci/pytorch/build.sh
Comment (jeanschmidt) [reply .ci/pytorch/build.sh]: 🤷🏻‍♂️

Changed so it runs for all...

**Comment by atalman** [comment]
@jeanschmidt to avoid code repetition in _linux-build-arc.yml we can try to rewrite some steps as actions then both _linux-build and _linux-build-arc can use same action

**Review by malfet — APPROVED** [review]
Thank you for getting rid of TODO and adding an explanation why separate .yml file is needed

**Pull request closed without merging**
